### PR TITLE
[UWP] Validate TrackEvent parameters

### PR DESCRIPTION
**Pull request by MatkovIvan**

### CODE REVIEW

**Comment by guperrot** [comment]
@achocron cannot get this PR to trigger VSTS build, please look.

**Comment by guperrot** [comment]
@MatkovIvan turns out PRs are not built by VSTS (only pushes) so we are enforcing up-to-date branch check on protected branches.

File: SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs [hunk 1]
```diff
@@ -114,6 +129,17 @@ private void InstanceTrackEvent(string name, IDictionary<string, string> propert
...
                 return;
             }
+            if (string.IsNullOrEmpty(name))
+            {
+                MobileCenterLog.Error(LogTag, "Name cannot be null or empty.");
+                return;
+            }
+            if (name.Length > MaxEventNameLength)
+            {
+                MobileCenterLog.Error(LogTag, "Name cannot be longer then " + MaxEventNameLength + " characters.");
```
Comment (guperrot) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: Messages should be synced with Android.
See https://github.com/Microsoft/mobile-center-sdk-android/pull/391#pullrequestreview-30926545

File: SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs [hunk 2]
```diff
@@ -154,6 +171,71 @@ private ISessionTracker CreateSessionTracker(IChannelGroup channelGroup, IChanne
...
+        /// Validates properties.
+        /// </summary>
+        /// <param name="properties">Properties collection to validate.</param>
+        /// <param name="logName">Log name.</param>
+        /// <param name="logType">Log type.</param>
+        /// <returns>Valid properties collection with maximum size of 5</returns>
+        private IDictionary<string, string> ValidateProperties(IDictionary<string, string> properties, string logName, string logType)
+        {
+            if (properties == null)
+                return null;
```
Comment (dhei) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: >return null; [](start = 16, length = 12)

nitpick: Omitting curly bracket in one-line return statement is error-prone. See microsoft stylecop SA1503 ([http://stylecop.soyuz5.com/sa1503.html](http://stylecop.soyuz5.com/sa1503.html))

File: SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs [hunk 3]
```diff
@@ -154,6 +171,71 @@ private ISessionTracker CreateSessionTracker(IChannelGroup channelGroup, IChanne
...
+            var result = new Dictionary<string, string>();
+            foreach (var property in properties)
+            {
+                if (result.Count >= MaxEventProperties)
```
Comment (dhei) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: nitpick: I personally feel 'break' and 'continue' in the for loop less readable compared to nested if-else's, would this make it more readable?

```C#
if (string.IsNullOrEmpty(property.Key))
{
	MobileCenterLog.Error(LogTag, string.Format("{0} '{1}' : a property key cannot be null or empty. Property will be skipped.", logType, logName));	
}
else if (property.Key.Length > MaxEventPropertyKeyLength) 
{
	MobileCenterLog.Error(LogTag, string.Format("{0} '{1}' : property '{2}' : property key length cannot be longer than {3} characters. Property '{2}' will be skipped.", logType, logName, property.Key, MaxEventPropertyKeyLength));
} 
else if (property.Value == null) 
{
	MobileCenterLog.Error(LogTag, string.Format("{0} '{1}' : property '{2}' : property value cannot be null. Property '{2}' will be skipped.", logType, logName, property.Key));
} 
else if (property.Value.Length > MaxEventPropertyValueLength) 
{
	MobileCenterLog.Error(LogTag, string.Format("{0} '{1}' : property '{2}' : property value cannot be longer than {3} characters. Property '{2}' will be skipped.", logType, logName, property.Key, MaxEventPropertyValueLength));
}
else if (result.Count < MaxEventProperties)
{
	result.Add(property.Key, property.Value);	
}
else 
{
	MobileCenterLog.Error(LogTag, string.Format("{0} '{1}' : properties cannot contain more than {2} items. Skipping other properties.", logType, logName, MaxEventProperties));
}
```

Comment (MatkovIvan) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: @dhei I'm OK to remove `continue`, but I disagree with your code. This variant will repeat "properties cannot contain more" log until for end. One log is enough.

Comment (dhei) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: @MatkovIvan Yes you are right, I missed that, good catch. 👍

**Comment by guperrot** [comment]
VSTS does not handle fork and PRs so we manually kicked off the build from pull/174/merge.
Our plan is to pilot that from bitrise to get proper github status reporting.

**Comment by guperrot** [comment]
VSTS build ok.
Manual testing ok.

File: SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs [hunk 4]
```diff
@@ -56,8 +61,14 @@ public static bool Enabled
         }
 
         /// <summary>
-        ///     Track a custom event.
+        ///     Track a custom event with name and optional properties.
         /// </summary>
+        /// <remarks>
+        ///     The name parameter can not be null or empty.Maximum allowed length = 256.
```
Comment (guperrot) [inline SDK/MobileCenterAnalytics/VSMC.Analytics.Windows.Shared/Analytics.cs]: Can you copy the Javadoc wording from Android SDK before we merge?

**Pull request merged**